docs/saved_objects/saved_object_access_control_design_proposal.md

```diff
@@ -0,0 +1,216 @@
+* library_read
+* library_write (library_read_and_write)
+
+the management permission allows the principal to update the workspace settings, like enabling or disabling features in the workspace, as well as updating the ACLs for the workspace, such as granting permissions of other principals to access this workspace. *library_read* and *library_write* permission will be used to control the user/group permissions on the saved objects in the workspace.
+
+With this permission definition
+
+* a user is the workspace admin if they have management permission of the workspace
+* a user can access/use a workspace if they have either library_read, or library_write permission or management permission of the workspace
+* a user can view a saved object if 
+    * they have read or write permission of the saved object, or 
+    * they have library_read or library_write or admin permission of one of the workspace that saved object is associated to.
+* a user can edit a saved object if
+    * they have write permission of the saved object, or
+    * they have library_write or admin permission of one of the workspace that saved object is associated to
+* a user can create saved object in a workspace if they have library_write permission of the workspace
+
+
+#### ACL
+
+The ACLs can be represented as following:
+
+```
+{
+  "permissions": {
+    "<permission_type_1>": [
+       "<principal_1>",
+       "<principal_2>"
+    ],
+    "<permission_type_2>": [
+       "<principal_3>",
+       "<principal_4>"
+    ],
+  } 
+}
+```
+
+We can also allow wildcard `*` to indicate any authenticated identity in the ACL.
+
+For example, if we want to allow finance_manager group to manage the workspace, and allow *finance_analyst* groups to build dashboards in the workspace, the workspace ACLs may look like:
+
+```
+{
+  "permissions": {
+    "management": [
+       "group/finance_manager"
+    ],
+    "library_write": [
+       "group/finance_analyst"
+    ],
+  } 
+}
+```
+
+While if we want to allow user-1 to modify a saved object, and one can only view a saved object, the ACL of that common saved object will look like:
+
+```
+{
+  "permissions": {
+    "read": [
+       "*"
+    ],
+    "library_write": [
+       "user/user-1"
+    ],
+  }
+}
+```
+
+
+ACLs can be considered as sub-resource or attributes of a OpenSearch Dashboards data record, so users who have the write permission to the data record can modify the ACL.
+
+**Orphan objects**
+This means a user can remove themselves from the ACL of a saved object. Once a user self-remove from a saved object’s ACL, other user having write permission can add the user back if needed. We can add additional checks to make sure any saved object will at least have some identities in the ACL to avoid orphan objects. However, in the case that the only owner of the object get removed from the user directory, which caused the objects becoming orphan, we will need super-admin to get involved to deal with the orphan data records.
```

Comment:
Requiring the intervention of a super-admin for orphaned objects seems like its going to cause maintenance/operation issues, can we alter the design to handle/avoid these scenarios?

Reply:
Objects not associate with workspace and do not have ACL defined, such objects can be considered "public", I think such objects won't require permission check. These saved objects behave like they are nowadays.